### ISSUE TITLE
[Feature] Support for tick marks instead of fully circled bubbles

### ISSUE THREAD

**Issue by Udayraj123** [Udayraj123/OMRChecker#167]
**Is your feature request related to a problem? Please describe.**
It'll be good to have support for tick marks/cross marks instead of fully circled bubbles.

**Describe the solution you'd like**
- Traverse each bubble point just like in the current algorithm
- Instead of calculating the mean pixel value for a bubble, use a different algorithm to accurately detect a cross or a tick mark (open to ideas)
- The algorithm should be able to differentiate between a slightly shifted bubble boundary(checkbox walls) vs a tick mark (no false positives) 
<img width="504" alt="image" src="https://github.com/Udayraj123/OMRChecker/assets/16881051/52c4dfcf-bbe3-4f67-a663-5ba497834669">

- Return a fixed value for that detected bubble so that it works with existing algorithm too.
```python
for pt in field_block_bubbles:
    # shifted
    x, y = (pt.x + field_block.shift, pt.y)
    rect = [y, y + box_h, x, x + box_w]
    q_strip_vals.append(
        # cv2.mean(img[rect[0] : rect[1], rect[2] : rect[3]])[0]
        detectCross(img, rect) ? 0 : 255
    )
```

**Additional context**
- The code comment [here](https://github.com/Udayraj123/OMRChecker/blob/f9f6dc3603738474bd29b79b4b1600c610d21120/src/core.py#L220) is a good starting point (perhaps the solution to this entire problem)
- [This stackoverflow discussion](https://stackoverflow.com/questions/72147441/how-to-detect-checkboxes-by-removing-noise-using-python-opencv)
- Can explore [cv2. connectedComponentsWithStats](https://loichau997.medium.com/apply-computer-vision-on-the-questionaire-image-to-detect-ticked-checkboxes-646e9245d293)

**Comment by Udayraj123** [comment]
Hello @alt-shreya, feel free to propose a basic solution/approach. Then I will assign the issue to you :)
You may also discuss it on discord if needed

**Comment by alt-shreya** [comment]
@Udayraj123 I went through the resources you had linked, and found the StackOverflow discussion very interesting.
Building upon the approach highlighted by [nathancy](https://stackoverflow.com/users/11162165/nathancy), a cross could be detected in the same way that an X is detected in handwriting recognition (2 diagonal lines of similar length and an intersection point), and a starting point to detecting tick marks could be to detect two lines inside the contour, in which the length of one line is shorter than the other

**Comment by Udayraj123** [comment]
Hey @alt-shreya, thanks for taking forward the discussion.

> cross could be detected in the same way that an X is detected in handwriting recognition 
Are you thinking of exploring using an ML algorithm for this?

> to detect two lines inside the contour, in which the length of one line is shorter than the other
Okay, but anything on distinguishing it from bubble boundary? If the box gets shifted left or right a little bit, the contour approach will probably get disturbed. Can you try something on the  shape approximation as done in the Stackoverflow discussion?

**Comment by Udayraj123** [comment]
> Could you give me an example of when it might shift?
I don't have images with tick marks handy, but I think the following will suffice.
So the shifting problem exists for any OMR sheets(even without tick marks). Usually happens in the bottom and bottom-right sections due to scanning/printing skews.
For filled bubbles we use the mean value for the whole box which usually works, so we ignore any slight shifts in the image. 

Here's a few examples for the general shifting problem -
From [template alignment](https://github.com/Udayraj123/OMRChecker/wiki/Rich-Visuals#template-layout-alignment) - 

<img width="527" alt="image" src="https://github.com/Udayraj123/OMRChecker/assets/16881051/91d1bdc5-6e8e-4136-bd9f-910080c194b8">

Try to imagine tick marks instead of the bubbles in above image.

The mean value approach also sometimes fails for filled bubbles case like this (false positive) -
<img width="489" alt="image" src="https://github.com/Udayraj123/OMRChecker/assets/16881051/cf1893aa-d390-4363-a900-d3b0d7a9ff4e">

**Comment by Udayraj123** [comment]
Here's a sample you can try this out with(will need to create a small template.json) - 

| Original | Shifted |
|----|----|
| <img width="126" alt="image" src="https://github.com/Udayraj123/OMRChecker/assets/16881051/7cc7d427-5a57-492a-bf2f-38802d764b04"> | <img width="129" alt="image" src="https://github.com/Udayraj123/OMRChecker/assets/16881051/e1914b7a-d9b0-46fb-883f-102a593caf71"> |

I've manually added a slight shift in the images.

Note that you will need a larger bubbleDimension in the template since the tick marks are overflowing the expected box.
Let me know if you need help with creating the initial template.json

**Comment by chirag4862** [comment]
I would Like to work on the solution for this if it is still open and not being worked on already

**Comment by Udayraj123** [comment]
Hey @chirag4862, this issue is labelled as intermediate and can be picked up after getting familiar with OMRChecker through any other issues first as this will take some time  Let's get you familiar with submitting your first sample!
